source/material_model/simple_nonlinear.cc

```diff
@@ -0,0 +1,435 @@
+                  // creep (where n_diff=1) viscosities are stress and strain-rate independent, so the calculation
+                  // of compositional field viscosities is consistent with any averaging scheme.
+                  // TODO: Mailed Bob and asked why the averaging should be arithmetic. Bob replyed that this has to
+                  //       do with effective medium theory. Have to look into this a bit more.
+                  const double stress_exponent_inv = 1/stress_exponent[c];
+                  composition_viscosities[c] = std::max(std::min(std::pow(prefactor[c],-stress_exponent_inv) * std::pow(edot_ii,stress_exponent_inv-1), max_visc[c]), min_visc[c]);
+                  Assert(dealii::numbers::is_finite(composition_viscosities[c]),ExcMessage ("Error: Viscosity is not finite."));
+                  if (derivatives != NULL)
+                    {
+                      if (edot_ii_strict > min_strain_rate[c] * min_strain_rate[c] && composition_viscosities[c] < max_visc[c] && composition_viscosities[c] > min_visc[c])
+                        {
+                          //strictly speaking the derivative is this: 0.5 * ((1/stress_exponent)-1) * std::pow(2,2) * out.viscosities[i] * (1/(edot_ii*edot_ii)) * deviator(in.strain_rate[i])
+                          composition_viscosities_derivatives[c] = 2 * (stress_exponent_inv-1) * composition_viscosities[c] * (1/(edot_ii*edot_ii)) * deviator(in.strain_rate[i]);
+                          //std::cout << composition_viscosities_derivatives[i] << ", dev strt:" << deviator(in.strain_rate[i]) << std::endl;
+                        }
+                      else
+                        {
+                          composition_viscosities_derivatives[c] = 0;
+                          //std::cout << "Set " << i << " to 0" << std::endl;
+                        }
+                    }
+                }
+
+              out.viscosities[i] = Utilities::weighted_p_norm_average(volume_fractions, composition_viscosities, viscosity_averaging_p);
+              Assert(dealii::numbers::is_finite(out.viscosities[i]),ExcMessage ("Error: Averaged viscosity is not finite."));
+
+              //for(int ccc = 0; ccc < in.composition[i].size(); ccc++)
+              //  std::cout << " comp " << ccc << " = " << in.composition[i][ccc] << std::endl;
+              if (derivatives != NULL)
+                {
+                  derivatives->viscosity_derivative_wrt_strain_rate[i] = Utilities::derivative_of_weighted_p_norm_average(out.viscosities[i],volume_fractions, composition_viscosities, composition_viscosities_derivatives, viscosity_averaging_p);
+
+#ifdef DEBUG
+                  for (int x = 0; x < dim; x++)
+                    for (int y = 0; y < dim; y++)
+                      if (!dealii::numbers::is_finite(derivatives->viscosity_derivative_wrt_strain_rate[i][x][y]))
+                        std::cout << "Error: Averaged viscosity to strain-rate devrivative is not finite." << std::endl;
```

Comment:
convert this into an `Assert` -- nothing good can happen if the condition is not true.

same below. You can then remove the `#ifdef DEBUG` as well